internal/pkg/aws/codepipeline/codepipeline_test.go

```diff
@@ -505,3 +505,180 @@ func TestCodePipeline_GetPipelineState(t *testing.T) {
+			callMocks: func(m codepipelineMocks) {
+				m.cp.EXPECT().ListPipelineExecutions(
+					&codepipeline.ListPipelineExecutionsInput{
+						MaxResults:   aws.Int64(1),
+						PipelineName: aws.String(mockPipelineName)}).Return(nil, mockErr)
+			},
+			expectedOut:   "",
+			expectedError: fmt.Errorf("list pipeline execution for pipeline-dinder-badgoose-repo: some error"),
+		},
+		"errors if no pipeline execution IDs are returned": {
+			callMocks: func(m codepipelineMocks) {
+				m.cp.EXPECT().ListPipelineExecutions(
+					&codepipeline.ListPipelineExecutionsInput{
+						MaxResults:   aws.Int64(1),
+						PipelineName: aws.String(mockPipelineName)}).Return(mockBadOutput, nil)
+			},
+			expectedOut:   "",
+			expectedError: fmt.Errorf("no pipeline execution IDs found"),
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			// GIVEN
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			mockClient := mocks.NewMockapi(ctrl)
+			mockrgClient := mocks.NewMockresourceGetter(ctrl)
+			mocks := codepipelineMocks{
+				cp: mockClient,
+				rg: mockrgClient,
+			}
+			tc.callMocks(mocks)
+
+			cp := CodePipeline{
+				client:   mockClient,
+				rgClient: mockrgClient,
+			}
+
+			// WHEN
+			actualOut, actualErr := cp.pipelineExecutionID(mockPipelineName)
+
+			// THEN
+			if actualErr != nil {
+				require.EqualError(t, tc.expectedError, actualErr.Error())
+			} else {
+				require.Equal(t, tc.expectedOut, actualOut)
+			}
+		})
+	}
+}
+
+func TestCodePipeline_RetryStageExecution(t *testing.T) {
+	mockPipelineName := "pipeline-dinder-badgoose-repo"
+	mockStageName := "Source"
+	failedActions := codepipeline.StageRetryModeFailedActions
+	mockPipelineExecutionID := aws.String("12345678-fake-exec-utio-nid987654321")
+	mockErr := errors.New("some error")
+	mockOutput := &codepipeline.RetryStageExecutionOutput{
+		PipelineExecutionId: aws.String("12345678-fake-exec-utio-nid987654321"),
+	}
+
+	tests := map[string]struct {
+		callMocks     func(m codepipelineMocks)
+		expectedOut   *string
+		expectedError error
+	}{
+		"happy path": {
+			callMocks: func(m codepipelineMocks) {
+				m.cp.EXPECT().ListPipelineExecutions(
+					&codepipeline.ListPipelineExecutionsInput{
+						MaxResults:   aws.Int64(1),
+						PipelineName: aws.String(mockPipelineName)}).Return(&codepipeline.ListPipelineExecutionsOutput{
+					PipelineExecutionSummaries: []*codepipeline.PipelineExecutionSummary{
+						{
+							PipelineExecutionId: aws.String("12345678-fake-exec-utio-nid987654321"),
+						},
+					},
+				}, nil)
+				m.cp.EXPECT().RetryStageExecution(
+					&codepipeline.RetryStageExecutionInput{
+						PipelineExecutionId: mockPipelineExecutionID,
+						PipelineName:        aws.String(mockPipelineName),
+						RetryMode:           aws.String(failedActions),
+						StageName:           aws.String(mockStageName),
+					}).Return(mockOutput, nil)
+			},
+			expectedOut: nil,
+		},
+		"should return error from api client": {
```

Comment:
Another interesting testcase to add would be to make sure we return `nil` if the stage is no longer retryable!

In general, with our unit tests we want to aim at testing _behaviors_ so our test case names should be things like: 
- "returns nil when stage is successfully retried"
- "returns nil when stage status is not failed"
- "returns a wrapped error on unexpected codepipeline error"